register-machines/src/main.rs

```diff
@@ -8,44 +8,114 @@ type State = (Label, HashMap<Register, u128>);
 enum Instruction {
     Add(Register, Label),
     Sub(Register, Label, Label),
-    Halt
+    Halt,
 }
-
 use Instruction::*;
+
+#[allow(dead_code)]
 fn eval_program(program: &[Instruction], state: &State) -> State {
-    unimplemented!();
+    let (mut label, mut regs) = state.clone();
+    while (0..program.len()).contains(&label) {
+        match program[label] {
+            Add(reg, new_label) => {
+                *regs.entry(reg).or_default() += 1;
+                label = new_label;
+            }
+            Sub(reg, new_label, other_label) => {
+                let register = regs.entry(reg).or_default();
+                label = if *register > 0 {
+                    *register -= 1;
+                    new_label
+                } else {
+                    other_label
+                }
+            }
+            Halt => break,
+        };
+    }
+    (label, regs)
 }
 // <<x,y>> = (2^x)*(2y+1)
+#[allow(dead_code)]
 fn encode_pair1(x: u128, y: u128) -> u128 {
-    unimplemented!();
+    (1 << x) * (2 * y + 1)
 }
 // <x,y> = (2^x)*(2y+1)-1
+#[allow(dead_code)]
 fn encode_pair2(x: u128, y: u128) -> u128 {
-    unimplemented!();
+    encode_pair1(x, y) - 1
 }
+#[allow(dead_code)]
 fn encode_list_to_godel(l: &[u128]) -> u128 {
-    unimplemented!();
+    match l {
+        [] => 0,
+        [x, l @ ..] => encode_pair1(*x, encode_list_to_godel(l)),
+    }
 }
+#[allow(dead_code)]
 fn encode_program_to_list(program: &[Instruction]) -> Vec<u128> {
-    unimplemented!();
+    program
+        .iter()
+        .map(|instruction| match instruction {
+            Add(reg, label) => encode_pair1(2 * reg, *label as u128),
+            Sub(reg, label, other_label) => encode_pair1(
+                2 * reg + 1,
+                encode_pair2(*label as u128, *other_label as u128),
+            ),
+            Halt => 0,
+        })
+        .collect()
 }
+
 // a = (2^x)*(2y+1)
+#[allow(dead_code)]
 fn decode_pair1(a: u128) -> (u128, u128) {
-    unimplemented!();
+    let x = a.trailing_zeros() as u128;
+    let y = a >> (x + 1);
+    (x, y)
 }
```

Comment:
Now `u128::trailing_zeros(u128)` is actually something that I thought definitely existed, but completely forgot to look for before implementing my horribly inefficient solution (by comparison). So well done!